### ISSUE TITLE
Turn stdarch into a subtree

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/stdarch#1655]
stdarch is one of the few remaining big submodules in the rustc repo. And it's painful -- stdarch depends on a lot of very internal rustc implementation details, but refactoring / improving those details always requires a complicated multi-stage process to ensure stdarch keeps building. It would be so much easier if the two could be updated in lock-step, but with a submodule that's not very practical. I ran into this twice just recently: as part of https://github.com/rust-lang/rust/issues/128738, and as part of https://github.com/rust-lang/rust/pull/131349. To ensure our continued ability of maintaining and improving rustc, we should fix this situation.

Miri, clippy, portable-simd, RA are all using subtrees, using either `git subtree` or [josh](https://github.com/josh-project/josh/). `git subtree` has some bad interacts with `git blame` due to the way it mirrors the changes between the repo. josh has the downside that many of the merge commits from the rustc repo end up being mirrored into the subrepo (see https://github.com/josh-project/josh/issues/1328). I don't care very strongly which is used, though the general trend has been away from `git subtree` and towards `josh` (RA recently migrated, clippy [decided to migrate](https://github.com/rust-lang/rust-clippy/pull/12759) but the effort seems currently stalled). If you pick `josh`, I can help with the migration. (I have some migration notes [here](https://hackmd.io/7pOuxnkdQDaL1Y1FQr65xg) but those are for a migration from `git subtree` to `josh`. Migrating from submodules will look a bit different.) You'll likely want some maintainer scripts in the stdarch repo to simplify pushing and pulling changes from and to the rustc repo; `cargo xtask` is probably the easiest way to set that up (that's what RA uses).

Last time I brought this up with @Amanieu, they were concerned about testing changes that land via rustc, since not the full stdarch test suite runs in rustc CI. I don't know how much testing in rustc CI is required to alleviate these concerns -- would be good to get a clear list of requirements. And then I hope someone can get motivated to work on this so that stdarch can stop being a roadblock for rustc evolution. :)

**Comment by lu-zero** [comment]
I think the main issue is writing the logic that triggers the stdarch CI bits only when something is changed on here.

`git subtree` seems available out of box, not sure how bad is the `git blame` issue or how difficult is to setup josh w/out docker so I cannot tell which is nicer to use. I noticed that [git-subrepo](https://github.com/ingydotnet/git-subrepo) seems available.

**Comment by RalfJung** [comment]
> I think the main issue is writing the logic that triggers the stdarch CI bits only when something is changed on here.

Generally we don't do this in rust-lang CI since it is rather fragile -- it is easy to miss some file that can make tests fail, and then it can be tricky to debug when a failure *actually* got introduced. CI should be stateless -- it should ensure that the repo satisfies certain checks, in a repeatable way that does not depend on past history. Miri and clippy for example get tested on rustc CI each time, not just when their subtrees change.

> how difficult is to setup josh w/out docker

I use josh without docker, `cargo install` is all it takes.

> git subtree seems available out of box

AFAIK one has to use a patched version to make it usable with large repos like rustc, otherwise it is way too slow.

> I noticed that [git-subrepo](https://github.com/ingydotnet/git-subrepo) seems available.

I have not tried this one. It seems to be squashing the subrepo history in the main repo into single large commits, whereas `git subtree` and `josh` both preserve individual subrepo commits.

**Comment by RalfJung** [comment]
Cc @calebzulawski @flip1995 @lnicola (for portable-simd, clippy, RA) in case they have anything to add to the choice of tool here.

**Comment by calebzulawski** [comment]
My only comment is that I often find myself deleting the branch and recreating it often when something goes a little awry with git-subtree. I haven't tried josh yet but it seems possibly easier to use.

**Comment by lu-zero** [comment]
Then I guess josh is a winner :)

**Comment by RalfJung** [comment]
This is still causing regular pain during compiler work, e.g. when cleaning up intrinsics in https://github.com/rust-lang/rust/pull/136543.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/issues/132735 is also running into this.

Can we make some sort of concrete plan here, so that stdarch stops being a constant thorn in the side of compiler contributors?

**Comment by lu-zero** [comment]
I guess the main question pending is who's doing the conversion and secondarily who's going to write the documentation so contributors won't get lost :)

**Comment by RalfJung** [comment]
If it's going to be done with josh-proxy, I can help, but the maintainers need to be closely involved to ensure this works long-term. rustc-dev-guide was recently migrated from a submodule to a josh-subtree so that could be used as a template (and it points out some of the potential pitfalls).
Cc @Kobzol

**Comment by Kobzol** [comment]
Sure, I'm willing to do the migration! I'll have to check stdarch's CI though, I remember it being a bit complex.

Do you think that it is necessary to run the full CI in rust-lang/rust, or could we get away with a subset that checks the most critical stuff, and then deal with the rest during pulls?

**Comment by Amanieu** [comment]
I believe the best plan for the long term is to fully migrate stdarch into rust-lang/rust and archive this repository. `stdarch` is only a separate repo for historical reason since it was originally an independent crate. Today `stdarch` is deeply tied to compiler and standard library internals so this no longer makes sense. Additionally it would help to run stdarch's testsuite on rustc changes since the tests frequently break on rustc changes, especially LLVM upgrades.

As such I think the best plan for a migration is:
1) Make rust-lang/rust CI run the stdarch testsuite, while keeping it as a - submodule. This would involve running the tests for all architectures under a single runner since the total runtime roughly matches a single rust-lang/rust runner.
2) Replace the submodule in rust-lang/rust with josh. Encourage stdarch development to happen on rust-lang/rust instead of stdarch.
3) Once all PRs and issues have been merged and/or migrated, the stdarch repository can be archived.

**Comment by RalfJung** [comment]
That would also work, of course -- and I agree re: all the implementation details.

Is it worth doing the josh setup if in the end we'll want to retire the separate repo anyway?

**Comment by Kobzol** [comment]
I'm fully in favor of moving stdarch into rust-lang/rust!

> Make rust-lang/rust CI run the stdarch testsuite, while keeping it as a - submodule. This would involve running the tests for all architectures under a single runner since the total runtime roughly matches a single rust-lang/rust runner.

That would require cross-compilation, I assume. Is stdarch configured in a way where this is easy to do?

I was more thinking about the opposite direction, i.e. run stdarch tests for architecture X in an existing rustc runner that already runs rustc (and other) tests for architecture X.

**Comment by Amanieu** [comment]
In theory it should just be a matter of executing all the `run-docker.sh` scripts in sequence on an appropriate runner: each docker container is set up for cross-compilation for the appropriate target and will run tests under qemu.

We have a handful of windows/mac targets which use plain `run.sh` without docker, those could be attached to existing runners.

**Comment by Kobzol** [comment]
Hmm, it's not that simple, I think. Linux jobs in r-l/r already run inside Docker, and tackling another Docker after tham would have a lot of complexity, I don't think we will want to do that. So we'll need to transplant the stdarch jobs into the rustc Docker containers somehow. Furthermore, the stdarch test suite will somehow need to be reimplemented within bootstrap, so it can be executed locally, and not just on CI. I'll take a look at how it could be done.

**Comment by BLANKatGITHUB** [comment]
hey @RalfJung  I was wondering as mentioned in issue the removal of [this](https://github.com/rust-lang/rust/blob/eca17022ef267f9ed87ba7d22755a1bfe0082e80/compiler/rustc_abi/src/extern_abi/mod.rs#L203-L206) well I couldn't find this file on main branch am I supposed to change it on master branch or a particular one?

**Comment by BLANKatGITHUB** [comment]
hey @RalfJung I was wondering as mentioned in issue the removal of [this](https://github.com/rust-lang/rust/blob/eca17022ef267f9ed87ba7d22755a1bfe0082e80/compiler/rustc_abi/src/extern_abi/mod.rs#L203-L206) well I couldn't find this file on main branch am I supposed to change it on master branch or a particular one?

**Comment by RalfJung** [comment]
(An off-topic post got removed, not sure by whom)

**Comment by Kobzol** [comment]
Migrating the test suite is... non-trivial. But we could turn stdarch into a (Josh) subtree anyway, and run at least some smoke tests on rust-lang/rust, in order to make these transitions easier. The full test suite would still run only in this repo.

**Comment by RalfJung** [comment]
Yeah, this could be paired with a cronjob that syncs and runs the full test suite against the latest rustc every morning. I guess the question is whether that's acceptable for @Amanieu.

**Comment by Amanieu** [comment]
Considering the amount of trouble this is causing, I'd be OK with just setting up a subtree now and figuring out the CI situation later.

**Comment by RalfJung** [comment]
Great, thanks :-)

**Comment by Kobzol** [comment]
Ok, since we're now also migrating compiler-builtins to a subtree, I think I'll revive the idea of creating some shared Josh sync scripts, and then I'll try to bootstrap the move into rust-lang/rust.

**Comment by RalfJung** [comment]
Happy to provide Josh advice where I can :)

**Comment by flip1995** [comment]
@Kobzol For a shared Josh sync script, also have a look at https://github.com/rust-lang/rust-clippy/pull/12759. We're doing some things slightly differently there, as we pin a nightly in a `rust-toolchain.toml` file in our repo, rather than pinning a commit hash and using RTIM, as `miri` does IIRC.

**Comment by RalfJung** [comment]
How do you ensure that a rustc-push uses the exact same commit as base as the one that the most recent rustc-pull pulled from? I would strongly recommend ensuring this invariant, it helps keep josh on the well-behaved paths.

**Comment by flip1995** [comment]
Link to our old [Zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/185405-t-compiler.2Frust-analyzer/topic/subtree.20issues/near/436872148) where we talked about this 😅 

Basically bumping the nightly in `rust-toolchain.toml` file to today's version, then using `rustc -vV` to get the commit of the nightly. That commit will then be used for both `pull` and `push`.

**Comment by Kobzol** [comment]
The move to Josh itself is implemented in https://github.com/rust-lang/rust/pull/141899.

**Comment by sayantn** [comment]
Just a small question, what will this entail for the contributors (sorry no experience with subtrees)? Will the PRs have to be sent to the rustc repo or this repo itself? And do we need to do something more when editing stdarch in a rustc PR?

**Comment by Kobzol** [comment]
Most changes should still be sent to this repository. Only when there is a change that also requires a corresponding rustc change (or if there is a rustc change that requires a stdarch change) should the PR be sent against rust-lang/rust.

>  And do we need to do something more when editing stdarch in a rustc PR?

I don't think so.

**Comment by sayantn** [comment]
Ah ok thanks, so no more 3-PR chains just to add some target features 🎉.

And one more question, will the changes to this repo be automatically synced to the rustc subtree, or do we have to do it manually like now?

**Comment by Kobzol** [comment]
https://github.com/rust-lang/rust/pull/141899 stdarch is a subtree now. There are still some things to do though (integrate it into bootstrap, send josh sync scripts to this repo, maybe setup some sync automation).

**Comment by RalfJung** [comment]
Amazing, thanks a lot :)

**Comment by Kobzol** [comment]
I think we can close this now, as stdarch is a subtree.

**Comment by RalfJung** [comment]
🎉 

Do we track somewhere what could/should be done on the CI side?

**Comment by Kobzol** [comment]
I don't think so, but also it's a bit unclear to me what we should do proactively. I think that we'll find out over time which kinds of failures we experience with stdarch updates (if any), and then we can add targeted CI tests to make sure that they don't happen again.
